[DSM] DDP-7556 Mercury Sequencing Orders

pegahtah opened the pull request:
Tickets in this PR 
https://broadinstitute.atlassian.net/browse/DDP-7556
https://broadinstitute.atlassian.net/browse/DDP-7604
https://broadinstitute.atlassian.net/browse/DDP-7608
https://broadinstitute.atlassian.net/browse/DDP-8050
https://broadinstitute.atlassian.net/browse/DDP-8115
https://broadinstitute.atlassian.net/browse/DDP-8116
https://broadinstitute.atlassian.net/browse/DDP-8217

pegahtah committed:
DDP-7556
* added the PR comments
* made the request to work with collaborator pt id as well

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into DDP-7556_Mercury_fixes

pegahtah committed:
DDP-7556
* fixed the checkstyle error

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into DDP-7556_Mercury_fixes

# Conflicts:
#	pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/DDPInstance.java

pegahtah committed:
DDP-7556
* moved try catch into `createAndPublishMessage` method and removed it from Dao

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into DDP-7556_Mercury_fixes

pegahtah committed:
DDP-7556
* moved the `createAllOrders` before the pubsub publishing to check for existing

pegahtah committed:
DDP-7556
* added the ability to always have test tissues with sample id

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into DDP-7556_Mercury_fixes

pegahtah committed:
Applying PR comments and fixes to the branch
*added mercuryOrderCreator to DDPInstanceDao
*fixed the SQL query for getting eligible tissues and samples
*fixed the json to have all the info in a `productOrder` object
*changed published json on pubsub to be base64
*added userId and ddpInstanceId columns to the backend to keep better track of the orders

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into DDP-7556_Mercury_fixes

# Conflicts:
#	pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/CreateClinicalDummyKitRoute.java

pegahtah committed:
Applying PR comments and fixes to the branch
*fixed merge conflicts

pegahtah committed:
Applying PR comments and fixes to the branch
*renamed Route to`PostMercuryOrderDummyRoute`
* removed `ddpParticipant Id` from `MercuryOrderDummyRequest`
* renamed `MercuryOrderRequest` class to `MercuryOrderDummyRequest`

pegahtah committed:
I added a Dao class `MercurySampleDao.java` for getting eligible samples for ordering per participant, and also the `GetMercuryEligibleSamplesRoute.java` route
 I added the instance and permission roles needed
 Also created the MercurySampleDto

pegahtah committed:
Adding collectionDate

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into DDP-7556_Mercury_backend

# Conflicts:
#	pepper-apis/dsm-core/src/main/resources/master-changelog.xml

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into DDP-7556_Mercury_backend

pegahtah committed:
Adding collectionDate
Fixing bug in search page to bring back all of the kits

pegahtah committed:
Added `PostMercuryOrderRoute` to place a sequencing order from frontend, also added methods to `TissueSMIDDao.java` and `KitRequestDao.java`
Added checks for getting only tissues that don't have a missing sm id in uss and scrolls in `MercurySampleDao.java`
Added `createBarcodes` function in `MercuryOrderUseCase.java` to get all the sm ids and the kit label
Added `MercuryOrderStatusListener` for listening to pubsub subscription

pegahtah committed:
* added ClinicalOrderDao for getting the Clinical Orders per realm for the Clinical Order's page
* added two columns of tissue_id and dsm_kit_Request_id instead of sample_id and changed the Dao and Dto classes accordingly
* Corrected bug in MercurySampleDao.java to avoid getting tissues more than once
* Added the ClinicalOrderUseCase.java to publish an ActionRequestMessage and also added the ActionRequestMessage
* I removed the base64 Encoding by Gp's request
* Added the functionality of getting orders in the GetMercuryEligibleSamplesRoute.java
* Added checks for access in all the routes
* Added the instance role in the DisplaySettingsRoute
* Added a changeset of `mercury_sequencing_further_changes` for changes to the table and foreign keys

pegahtah committed:
* Adding the MercuryOrderStatusListener to DSMServer

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into DDP-7556_Mercury_backend

pegahtah committed:
adding `sequencing_restriction` field

pegahtah committed:
removing the FON test route

pegahtah committed:
adding sequencingRestriction to the object

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into DDP-7556_Mercury_backend

# Conflicts:
#	pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/DSMServer.java
#	pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/mercury/MercuryOrderDao.java
#	pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dto/mercury/MercuryOrderDto.java
#	pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dto/mercury/MercuryOrderUseCase.java
#	pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/pubsub/MercuryOrderPublisher.java
#	pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/CreateClinicalDummyKitRoute.java
#	pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/mercury/PostMercuryOrderDummyRoute.java
#	pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/mercury/PostMercuryOrderRoute.java
#	pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/statics/DBConstants.java
#	pepper-apis/dsm-core/src/main/resources/master-changelog.xml

pegahtah committed:
added test for accessioning
added mappings for new fields

pegahtah committed:
added tracking times and checks for accessioning
added mappings for new fields

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into DDP-7556_Mercury_backend

pegahtah committed:
DDP-8217 adding the new columns in sm id table and updating them
DDP-8217 adding date_px ad collectionDate for sm ids
DDP-7556 DisplaySettingsRoute returns if the realm has sequencting

pegahtah committed:
DDP-8217 only adding sm ids with received_date (Accessioned) to the order

pegahtah committed:
DDP-7604 Since collection date field is added, we have to make sure dummy kits also have that

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into DDP-7556_Mercury_backend

gmakharat commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/mercury/MercuryOrderDao.java:
```diff
@@ -189,4 +195,44 @@ public void insertMercuryOrders(List<MercuryOrderDto> newOrders) {
+        return 0;
+    }
+
+    public SimpleResult create(MercuryOrderDto mercuryOrderDto, Connection conn) {
```
why it became necessary to pass `Connection` here?

gmakharat replied:
still actual question

SimoneMaiwald replied:
@pegahtah ?

pegahtah replied:
Ooh thanks for the reminder! It was part of the old way I made it in which I wanted to have a way to roll back the connection if pubsub message failed, but I decided not to have insertions dependent on pubsub messages

gmakharat commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dto/mercury/MercuryOrderUseCase.java:
```diff
@@ -29,4 +36,16 @@ public static List<MercuryOrderDto> createAllOrders(String[] barcodes, String dd
         return orders;
     }
+
+    public static ArrayList<String> createBarcodes(MercurySampleDto[] mercurySampleDtos, DDPInstanceDto ddpInstance) {
```
static methods break whole idea of Object Oriented Programming

SimoneMaiwald replied:
naming is not correct. we are not creating barcodes

gmakharat commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/mercury/GetMercuryEligibleSamplesRoute.java:
```diff
@@ -0,0 +1,69 @@
+        if (!UserUtil.checkUserAccess(realm, userId, "mercury_order_sequencing", null)) {
+            log.warn("User doesn't have access");
+            response.status(500);
+            return new Result(500, UserErrorMessages.NO_RIGHTS);
```
no need to return `Result` object anymore, just return `UserErrorMessages.NO_RIGHTS` directly.

gmakharat commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/mercury/GetMercuryEligibleSamplesRoute.java:
```diff
@@ -0,0 +1,69 @@
+            response.status(500);
+            return new Result(500, UserErrorMessages.NO_RIGHTS);
+        }
+        if (request.url().contains(RoutePath.MERCURY_SAMPLES_ROUTE)) {
```
I would create two separate route classes for samples and for orders.

pegahtah replied:
The other one is moved to another new Route

gmakharat commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/mercury/PostMercuryOrderDummyRoute.java:
```diff
@@ -45,6 +46,12 @@ public Object handle(Request request, Response response) {
+        if (!UserUtil.checkUserAccess(realm, userId, "mercury_order_sequencing", null)) {
+            log.warn("User doesn't have access");
+            response.status(500);
+            return new Result(500, UserErrorMessages.NO_RIGHTS);
```
the same about `Result` object.

gmakharat commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/mercury/PostMercuryOrderRoute.java:
```diff
@@ -0,0 +1,74 @@
+            realm = queryParams.get(RoutePath.REALM).value();
+        } else {
+            throw new RuntimeException("No realm query param was sent");
+        }
```
Optional:

String realm = Optional.ofNullable(queryParams.get(RoutePath.REALM).value())
              .orElseThrow(() -> new RuntimeException("No realm query param was sent"));

gmakharat commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/mercury/PostMercuryOrderRoute.java:
```diff
@@ -0,0 +1,74 @@
+            ddpParticipantId = queryParams.get(RoutePath.DDP_PARTICIPANT_ID).value();
+        } else {
+            throw new RuntimeException("No realm query param was sent");
+        }
```
Optiona:

the same can be applied to this like in previous comment.

gmakharat commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/mercury/PostMercuryOrderRoute.java:
```diff
@@ -0,0 +1,74 @@
+        if (!UserUtil.checkUserAccess(realm, userId, "mercury_order_sequencing", null)) {
+            log.warn("User doesn't have access");
+            response.status(500);
+            return new Result(500, UserErrorMessages.NO_RIGHTS);
```
same for `Result` object.

gmakharat commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/mercury/PostMercuryOrderRoute.java:
```diff
@@ -0,0 +1,74 @@
+        }
+        MercurySampleDto[] mercurySampleDtos = new Gson().fromJson(requestBody, MercurySampleDto[].class);
+        DDPInstanceDto ddpInstance = new DDPInstanceDao().getDDPInstanceByInstanceName(realm).orElseThrow();
+        if (ddpInstance == null) {
```
I think this check does not make sense, since `getDDPInstanceByInstanceName(realm).orElseThrow()` already throws an exception if realm does not exist.

gmakharat commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/mercury/PostMercuryOrderRoute.java:
```diff
@@ -0,0 +1,74 @@
+            return new Result(500, UserErrorMessages.CONTACT_DEVELOPER);
+        }
+        publishMessage(mercurySampleDtos, ddpInstance, userId, ddpParticipantId);
+        return new Result(200);
```
just set response to 200 and return StringUtils.EMPTY;

gmakharat commented on pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/mercury/MercuryEndpointTests.java:
```diff
@@ -0,0 +1,49 @@
+
+        accessionCommand = String.format(accessionCommand, bspToken, barcode);
+        try {
+            Process process = Runtime.getRuntime().exec(accessionCommand);
```
I think it worths to add some assertion checks here to check values returned by response, if I got correctly intent of this test is to check creation of clinical kits, right?

SimoneMaiwald commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/mercury/GetMercuryEligibleSamplesRoute.java:
```diff
@@ -0,0 +1,69 @@
+        }
+        String realm = queryParams.get(RoutePath.REALM).value();
+        if (!UserUtil.checkUserAccess(realm, userId, "mercury_order_sequencing", null)) {
+            log.warn("User doesn't have access");
```
add userid to the warning as well

SimoneMaiwald commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/mercury/GetMercuryEligibleSamplesRoute.java:
```diff
@@ -0,0 +1,69 @@
+            throw new RuntimeException("No realm query param was sent");
+        }
+        String realm = queryParams.get(RoutePath.REALM).value();
+        if (!UserUtil.checkUserAccess(realm, userId, "mercury_order_sequencing", null)) {
```
why are we not checking if userId matches?

SimoneMaiwald commented on pepper-apis/dsm-core/src/main/resources/liquibase/CMI/DDP-7556_mercury_sequencing_roles.xml:
```diff
@@ -0,0 +1,60 @@
+
+    <changeSet id="DDP-7556_mercury_roles" author="pegah">
+        <insert tableName="instance_role">
+            <column name="name" value="has_mercury_order_sequencing"></column>
```
has_clinical_kit

SimoneMaiwald commented on pepper-apis/dsm-core/src/main/resources/liquibase/CMI/DDP-7556_mercury_sequencing_roles.xml:
```diff
@@ -0,0 +1,60 @@
+            <column name="name" value="has_mercury_order_sequencing"></column>
+        </insert>
+        <insert tableName="access_role">
+            <column name="name" value="mercury_order_sequencing"></column>
```
kit_sequencing_order

SimoneMaiwald commented on pepper-apis/dsm-core/src/main/resources/liquibase/CMI/DDP-7556_mercury_sequencing_roles.xml:
```diff
@@ -0,0 +1,60 @@
+            <column name="collection_date" type="VARCHAR(45)"></column>
+        </addColumn>
+        <addColumn tableName="ddp_kit">
+            <column name="last_changed" type="BIGINT"></column>
```
BIGINT(20)

SimoneMaiwald commented on pepper-apis/dsm-core/src/main/resources/liquibase/CMI/DDP-7556_mercury_sequencing_roles.xml:
```diff
@@ -0,0 +1,60 @@
+        </insert>
+    </changeSet>
+    <changeSet id="DDP-7556_mercury_sequencing_order_table_id_changes" author="pegah">
+        <addColumn tableName="mercury_sequencing">
```
where is that table created?

SimoneMaiwald replied:
please rename the table to ddp_mercury_sequencing

pegahtah replied:
The table creation is already merged for gp testing, in changeset `7556_mercury_sequencing`

SimoneMaiwald commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/DisplaySettingsRoute.java:
```diff
@@ -136,6 +136,10 @@ public Object processRequest(Request request, Response response, String userId)
                 if (DDPInstanceDao.getRole(instance.getName(), DBConstants.SHOW_GROUP_FIELDS)) {
                     displaySettings.put("showGroupFields", true);
                 }
+                if (DDPInstanceDao.getRole(instance.getName(), DBConstants.HAS_ORDER_SEQUENCING)) {
```
Now we are calling too often to get a role!
Please 
1) add a method to get a List of roles for a given instance by name
2) change class to check if list contains the role

SimoneMaiwald commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/ClinicalKitsRoute.java:
```diff
@@ -57,7 +57,7 @@ private ClinicalKitDto getClinicalKit(String kitLabel, Optional<BSPKitDto> optio
         } else {
-            Optional<KitInfo> maybeKitInfo = GPReceivedKit.receiveKit(kitLabel, optionalBSPKitDto.get(), notificationUtil, mercury);
+            Optional<KitInfo> maybeKitInfo = GPReceivedKit.receiveKit(kitLabel, optionalBSPKitDto.get(), notificationUtil, MERCURY);
             KitInfo kitInfo = maybeKitInfo.get();
```
potential NPE?

SimoneMaiwald commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/ClinicalKitsRoute.java:
```diff
@@ -57,7 +57,7 @@ private ClinicalKitDto getClinicalKit(String kitLabel, Optional<BSPKitDto> optio
             //kit not found in ddp_kit table -> check tissue smi-ids
```
rename `getClinicalKit` to `getClinicalFFPEKit`

SimoneMaiwald commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/mercury/BaseMercuryStatusMessage.java:
```diff
@@ -0,0 +1,8 @@
+import lombok.Data;
+
+@Data
+public class BaseMercuryStatusMessage {
```
do we really need that message? are there more than just the MercuryStatusMessage ?

pegahtah replied:
The message is made in an order that looks like this 

```
Example of a order status return:
{
 “status”: {
"orderId":"Pepper-generated order id" 
"status": "Approved",
“details”: ”Successfully created PDO-123”,
“pdo”: “PDO-123”
}
}


```

SimoneMaiwald commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/gp/GPReceivedKit.java:
```diff
@@ -79,7 +80,7 @@ public static Optional<KitInfo> receiveKit(String kitLabel, BSPKitDto bspKitQuer
         return Optional.of(
                 new KitInfo(bspKitQueryResult.getBspCollection(), bspOrganism, "U", bspParticipantId, bspSampleId, bspMaterialType,
-                        bspReceptacleType, ddpInstance.getName(), bspKitQueryResult.getKitTypeName()));
+                        bspReceptacleType, ddpInstance.getName(), bspKitQueryResult.getKitTypeName(), collectionDate));
```
Is that change communicated between us and GP? BSP doesn't care for collection date!

pegahtah replied:
fixed, we are now returning another object for gp that doesn't have colelctionDate and KitInfo is used only when receiving for both BSP and MERCURY

SimoneMaiwald commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/gp/KitInfo.java:
```diff
@@ -19,9 +19,10 @@ public class KitInfo {
 
     public KitInfo(String sampleCollectionBarcode, int organismClassificationId, String gender, String bspParticipantId, String bspSampleId,
-                   String materialInfo, String receptacleName, String realm, String kitTypeName) {
+                   String materialInfo, String receptacleName, String realm, String kitTypeName, String collectionDate) {
```
Why is that change needed here?

pegahtah replied:
Because for clinical Kit route we also return KitInfo and we need to return collection date, I fixed it for BSP

SimoneMaiwald commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dto/mercury/ClinicalOrderUseCase.java:
```diff
@@ -0,0 +1,24 @@
+import org.broadinstitute.dsm.model.mercury.ActionRequestMessage;
+import org.broadinstitute.dsm.pubsub.MercuryOrderPublisher;
+
+@Slf4j
```
are we now using another logger?

SimoneMaiwald commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/mercury/MercurySampleDao.java:
```diff
@@ -0,0 +1,130 @@
+            } catch (SQLException e) {
+                dbVals.resultException = e;
+            }
+            try (PreparedStatement statement = conn.prepareStatement(SQL_GET_ELIGIBLE_SAMPLES)) {
```
split into separate methods

SimoneMaiwald commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/mercury/ClinicalOrderDao.java:
```diff
@@ -0,0 +1,113 @@
+        DDPInstanceDto ddpInstance = new DDPInstanceDao().getDDPInstanceByInstanceName(realm).orElseThrow();
+        HashMap<String, ClinicalOrderDto> map = new HashMap<>();
+        ArrayList<ClinicalOrderDto> noStatusOrders = new ArrayList<>();
+        SimpleResult results = inTransaction((conn) -> {
```
move into separate method

pegahtah replied:
done

SimoneMaiwald commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/ddp/kitrequest/KitRequestDao.java:
```diff
@@ -145,4 +150,32 @@ public String getKitRequestIdByBSPSampleId(String bspSampleId) {
         return (String) results.resultValue;
     }
+
+    public ArrayList<String> getKitLabelFromDsmKitRequestId(long dsmKitRequestId) {
```
why are we returning a list if a list would throw an error and never returned?

SimoneMaiwald commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java:
```diff
@@ -1468,9 +1475,9 @@ public static List<KitRequestShipping> findKitRequest(@NonNull String field, @No
                         if (Arrays.asList(realms).contains(realm)) { //only if user has right to see that realm
                             String ddpParticipantId = rs.getString(DBConstants.DDP_PARTICIPANT_ID);
-                            String bspParticipantId = rs.getString(DBConstants.COLLABORATOR_PARTICIPANT_ID);
+                            String bspSampleId = rs.getString(DBConstants.BSP_COLLABORATOR_SAMPLE_ID);
```
why are we changing this?

pegahtah replied:
because if participant has multiple samples then the key breaks

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into DDP-7556_Mercury_backend

# Conflicts:
#	pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/DSMServer.java
#	pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/mercury/MercuryOrderDao.java
#	pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/pubsub/MercuryOrderPublisher.java
#	pepper-apis/dsm-core/src/main/resources/master-changelog.xml

pegahtah committed:
DDP-7556
PR Comments

pegahtah committed:
DDP-7556
PR Comments

pegahtah committed:
DDP-7556
PR Comments

pegahtah committed:
Merge branch 'develop' into DDP-7556_Mercury_backend

gmakharat commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java:
```diff
@@ -284,6 +287,8 @@ public class KitRequestShipping extends KitRequest implements HasDdpInstanceId {
     @ColumnName(DBConstants.DDP_INSTANCE_ID)
     private long ddpInstanceId;
 
+    private String collectionDate;
```
as I got this field needs to be exported to ES, thus I believe we need to add `@ColumnName` annotation to make ES exporting work

gmakharat commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/ddp/instance/DDPInstanceDao.java:
```diff
@@ -113,6 +118,32 @@ public static boolean getRole(@NonNull String realm, @NonNull String role) {
         return (boolean) results.resultValue;
     }
 
+    public static HashSet<String> getInstanceRoles(@NonNull String realm) {
```
please make this method non static and return type more generic `Set<String>`

gmakharat commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/ddp/tissue/TissueSMIDDao.java:
```diff
@@ -226,4 +230,28 @@ public Map<String, List<SmId>> getSmIdsByParticipantByStudy(String instanceName)
         return smIds;
     }
+
+    public ArrayList<String> getSequencingSmIdsForTissue(long tissueId) {
```
It is always recommended to have return types more generic like `List<String>`

gmakharat commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/mercury/ClinicalOrderDao.java:
```diff
@@ -0,0 +1,113 @@
+        SimpleResult results = inTransaction((conn) -> {
+            SimpleResult dbVals = new SimpleResult();
+            try (PreparedStatement stmt = conn.prepareStatement(SQL_GET_ALL_ORDERS_FOR_REALM)) {
+                stmt.setInt(1, ddpInstance.getDdpInstanceId());
```
I would create inner select like `WHERE ms.ddp_instance_id = (SELECT ddp_instance_id FROM ddp_instance WHERE instance_name = ?)`, this would help to eliminate line 54.

gmakharat commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/mercury/ClinicalOrderDao.java:
```diff
@@ -0,0 +1,113 @@
+                return (int) (c2.orderDate - c1.orderDate);
+            }
+        });
+        clinicalOrderUseCase.publishStatusActionMessage(array, projectId, topicId);
```
not a good idea, `DAO` classes and their methods should have only one responsibility and it should be related only to CRUD operations, remove this operation from here and do it separately where you call `getOrdersForRealm`.

gmakharat commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/mercury/MercurySampleDao.java:
```diff
@@ -0,0 +1,155 @@
+        return Optional.empty();
+    }
+
+    public ArrayList<MercurySampleDto> findEligibleSamples(String ddpParticipantId, String realm) {
```
let's change to `List` return type.

gmakharat commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/mercury/MercurySampleDao.java:
```diff
@@ -0,0 +1,155 @@
+        return samples;
+    }
+
+    private ArrayList<MercurySampleDto> getEligibleKits(Connection conn, String ddpParticipantId, String realm) throws Exception {
```
let's change to `List` return type.

gmakharat commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/mercury/MercurySampleDao.java:
```diff
@@ -0,0 +1,155 @@
+        return samples;
+    }
+
+    private ArrayList<MercurySampleDto> getEligibleTissues(Connection conn, String ddpParticipantId, String realm) throws Exception {
```
let's change to `List` return type.

gmakharat commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dto/mercury/MercuryOrderUseCase.java:
```diff
@@ -29,4 +36,16 @@ public static List<MercuryOrderDto> createAllOrders(String[] barcodes, String dd
         return orders;
     }
+
+    public ArrayList<String> collectBarcodes(MercurySampleDto[] mercurySampleDtos) {
```
let's change to `List` return type.

gmakharat commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/pubsub/MercuryOrderStatusListener.java:
```diff
@@ -0,0 +1,64 @@
+            logger.info("Got STATUS message with Id: " + message.getMessageId());
+
+            try {
+                TransactionWrapper.inTransaction(conn -> {
```
Instead of having transaction here, why not to use dao class directly and you won't have to pass the connection as well

pegahtah committed:
DDP-7556 PR comments
* Changing the return type from `ArrayList` to `List` in a number of classes
* adding `@ColumnName(DBConstants.COLLECTION_DATE)` for CollectionDate in `KitRequestShipping`
* removing static from `getInstanceRoles` method
* removing typo in name `processOrderStatus()`
* moving `clinicalOrderUseCase.publishStatusActionMessage(orders, projectId, topicId);` to GetMercuryOrdersRoute.processRequest
* adding  check for Failed messages
* changing the thrown exception in `updateOrderStatus` to `DSMPubSubException`
* removing unnecessary `static` in `MercuryOrderPublisher`

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into DDP-7556_Mercury_backend

pegahtah committed:
DDP-7556
* removing unused fields

pegahtah commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/pubsub/MercuryOrderPublisher.java:
```diff
@@ -29,18 +29,18 @@
-    private static ParticipantDao participantDao;
+    private MercuryOrderDao mercuryOrderDao;
+    private ParticipantDao participantDao;
 
```
@bskinner I removed the statics you mentioned, Thanks for catching that!

bskinner replied:
Thank you!

pegahtah committed:
Merge branch 'develop' into DDP-7556_Mercury_backend

pegahtah committed:
Merge branch 'develop' into DDP-7556_Mercury_backend

pegahtah committed:
Merge branch 'develop' into DDP-7556_Mercury_backend

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into DDP-7556_Mercury_backend

# Conflicts:
#	pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/DSMServer.java

pegahtah committed:
merging develop and removing old `DownloadParticipantFileRoute`

devonbush reviewed (dismissed):
Approve assuming you'll add the extra assertions in a subsequent commit

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into DDP-7556_Mercury_backend

SimoneMaiwald commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/ddp/instance/DDPInstanceDao.java:
```diff
@@ -63,6 +64,10 @@ public class DDPInstanceDao implements Dao<DDPInstanceDto> {
                     + "AND realm.ddp_instance_id = main.ddp_instance_id) AS 'has_role', mr_attention_flag_d, tissue_attention_flag_d, "
                     + "auth0_token, notification_recipients FROM ddp_instance main "
                     + "WHERE is_active = 1";
+    private static final String SQL_SELECT_ROLES_FOR_INSTANCE =
```
if that query is not used anywhere else, please add the by_instance_name part as well

SimoneMaiwald commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/mercury/ClinicalOrderDao.java:
```diff
@@ -0,0 +1,106 @@
+        }
+        ArrayList<ClinicalOrderDto> array = new ArrayList(map.values());
+        log.info(String.format("Found %d orders for realm %s", array.size(), realm));
+        Collections.sort(array, new Comparator<ClinicalOrderDto>() {
```
could the sorting be done by sql already?

pegahtah committed:
Merge branch 'develop' into DDP-7556_Mercury_backend

pegahtah committed:
Merge branch 'DDP-7556_Mercury_backend' of github.com:broadinstitute/ddp-study-server into DDP-7556_Mercury_backend

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into DDP-7556_Mercury_backend

pegahtah commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/mercury/MercurySampleDao.java:
```diff
@@ -0,0 +1,156 @@
+                    + "left join (select max(order_date) as order_date, tissue_id as seqt "
+                    + "From ddp_mercury_sequencing where ddp_participant_id = ? group by (tissue_id)   ) as table2 "
+                    + "on table2.seqt = table1.tissue_id ";
+
```
@SimoneMaiwald see here

SimoneMaiwald replied:
in easy words: when is a tissue eligible?

pegahtah replied:
Tissue should have collection date as date_px, be received and sent to GP,  and have all sm ids received, also number of uss and scrolls numbers should match the number of those inserted ( like if they entered 20 uss, they should have entered all the 20 uss values)

pegahtah replied:
tissue is the complex one, I am doing all of that in sql now

pegahtah commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/mercury/MercurySampleDao.java:
```diff
@@ -0,0 +1,156 @@
+                    + "    LEFT JOIN (select max(order_date) AS order_date, dsm_kit_request_id "
+                    + "    FROM ddp_mercury_sequencing WHERE ddp_participant_id = ? GROUP BY (dsm_kit_request_id)  ) AS table2 "
+                    + "    ON table2.dsm_kit_request_id = table1.dsm_kit_request_id";
+
```
@SimoneMaiwald see here

SimoneMaiwald replied:
when is a sample eligible?

pegahtah replied:
sample should be received, not having sequencing restrictions set to research only

pegahtah committed:
DDP-7556 not passing connection to `MercuryOrderDao.create` method anymore
adding the `BY_INSTANCE_NAME` query extension to `SQL_SELECT_ROLES_FOR_INSTANCE` in `DDPInstanceDao`

pegahtah committed:
DDP-7556 updated mapping and created migrator

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into DDP-7556_Mercury_backend

pegahtah committed:
DDP-7556 added filtering ability for ClinicalOrders of a particiapnt

pegahtah committed:
DDP-7556 name change in mapping to `clinicalOrder`

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into DDP-7556_Mercury_backend

pegahtah committed:
DDP-7556
Adding the functionality to export to ES a new published order
Changing type of `orderId` to keyword

pegahtah committed:
DDP-7556 checkstyle fix

pegahtah committed:
DDP-7556
Exporting Status updates

pegahtah committed:
DDP-7556
fixed a bug in query

pegahtah committed:
DDP-7556
Exporting the two new fields

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into DDP-7556_Mercury_backend

# Conflicts:
#	pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java

pegahtah committed:
DDP-7556
Exporting Status updates

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into DDP-7556_Mercury_backend

devonbush commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/DSMServer.java:
```diff
@@ -737,6 +745,12 @@ private void setupPubSub(@NonNull Config cfg, NotificationUtil notificationUtil)
 
+        try {
+            MercuryOrderStatusListener.subscribeToOrderStatus(projectId, mercuryDsmSubscriptionId);
+        } catch (Exception e) {
```
curious why you're using printStackTrace(), do we not have a logger available in this class?

pegahtah replied:
Honestly I just followed the pattern here

devonbush commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/elastic/export/generate/PropertyInfo.java:
```diff
@@ -21,16 +22,20 @@
 public class PropertyInfo {
```
I know it wasn't created in this PR, but this class could use a top line comment -- "PropertyInfo" could mean anything :)

devonbush commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/elastic/migration/ClinicalOrderMigrator.java:
```diff
@@ -0,0 +1,20 @@
+import org.broadinstitute.dsm.db.dao.mercury.ClinicalOrderDao;
+import org.broadinstitute.dsm.statics.ESObjectConstants;
+
+public class ClinicalOrderMigrator extends BaseCollectionMigrator {
```
Could you add a top-line comment to this class?  Or better yet, add one to BaseCollectionMigrator BaseMigrator explaining what migrators are built to do?

devonbush commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/gp/BSPKitInfo.java:
```diff
@@ -0,0 +1,44 @@
+package org.broadinstitute.dsm.model.gp;
+
```
am I the only person who doesn't know what BSP stands for?

pegahtah replied:
I don't know too, but it is a server in GP I believe

devonbush commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/mercury/PostMercuryOrderRoute.java:
```diff
@@ -0,0 +1,64 @@
+import spark.Request;
+import spark.Response;
+
+@Slf4j
```
this could use a comment explaining what the route is built to do

devonbush commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/mercury/GetMercuryOrdersRoute.java:
```diff
@@ -0,0 +1,54 @@
+import spark.Request;
+import spark.Response;
+
+@Slf4j
```
this could use a comment explaining what this route is intended for

devonbush commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/mercury/GetMercuryOrdersRoute.java:
```diff
@@ -0,0 +1,54 @@
+            response.status(500);
+            return UserErrorMessages.NO_RIGHTS;
+        }
+        if (request.url().contains(RoutePath.GET_MERCURY_ORDERS_ROUTE)) {
```
this seems like an unusual `if` -- shouldn't it be DSMServer/the router's responsibility to check that?  How else would this method get invoked?

devonbush commented:
I don't know enough about the context of this change, so don't let my comments delay anything -- they are FYI only :)

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into DDP-7556_Mercury_backend

pegahtah committed:
Merge branch 'develop' into DDP-7556_Mercury_backend
